### PR TITLE
Fix allowUnmappedFields, mapUnmappedFieldAsString settings to be applied when parsing query string query

### LINKED ISSUE #12141: [BUG] Error not returned when creating a percolator document containing a field without mapping
### Describe the bug

If an asterisk is included in the value of the query field when creating a percolator document with query_string, query_shard_exception (No field mapping can be found for the field with name [{someField}]) is not returned even if there is no mapping of that field. And that field is not searchable.

### Related component

Indexing

### request
PUT /:index/_doc/:id
```js
{
  "query": {
    "query_string": {
      "query": "someField: *test",
      "default_operator": "AND",
      "default_field": "body"
    }
  }
}
```

then this percolator document is created successfully even if `someField` has no mapping in its index and `someField` is not searchable

### Expected behavior

When the request body is(no asterisk)
```js
{
  "query": {
    "query_string": {
      "query": "someField: test",
      "default_operator": "AND",
      "default_field": "body"
    }
  },
}
```

cluster returns `400 Bad Requeest` response with `query_shard_exception(reason: No field mapping can be found for the field with name [someField])` 

### Additional Details

**Plugins**
Please list all plugins currently enabled.

**Screenshots**

**Host/Environment (please complete the following information):**

**Additional context**

### index settings
- I don't think there is a relevant setting

```js
"settings": {
      "index": {
        "replication": {
          "type": "DOCUMENT"
        },
        "refresh_interval": "3s",
        "number_of_shards": "1",
        "provided_name": "{index name}",
        "creation_date": "1706062096956",
        "number_of_replicas": "3",
        "uuid": "Dur7_p13SgOIwl-3rHt_BQ",
        "version": {
          "created": "136327927"
        }
      }
    }
```

### PR DESCRIPTION
**Pull request by imyp92**
### Description
PR to solve problems where the allowUnmappedFields, mapUnmappedFieldAsString settings do not apply to certain query types when parsing query_string queries containing unmapped fields. I've modified to call the QueryShardContext.fieldMapper() method, which applies the allowUnmappedFields, mapUnmappedFieldAsString setting to the return of MapperService.fieldType() instead of calling and using the return of MapperService.fieldType() directly.

### Related Issues
Resolves #12141 

#### Further description of the issue
- In the process of parsing a query_string, the following types of queries do not throw exception of 'QueryShardException(this, "No field mapping can be found for the field with name [{}], name)' even if the allowUnmappedFields and mapUnmappedFieldAsString settings are set to false.
  - wildcard query
    - `unmappedField: *` or `unmappedField: *test`
  - `_exists_` query
    - `_exists_: unmappedField`
  - range query
    - `unmappedField: <100`
  - fuzzy query 
    - `unmappedField: test~`
  - prefix query
    - `unmappedField: test*`

- The default value of `allowUnmappedFields` is true, and the default value of `mapUnmappedFieldAsString` is false, so if those settings are not customized, there will be no issue with search, but when indexing percolator documents, the issue arises regardless of whether the setting is customized or not because the allowUnmappedFields field of context object is set to false in `PercolatorFieldMapper.parse()` by invoking `PercolatorFieldMapper.configureContext()`.
#### To Reproduce
1. Create a test index
```sh
curl --location --request PUT 'localhost:9200/test' \
--header 'Content-Type: application/json' \
--data '{
  "settings": {      
    "index": {
      "query": {
        "parse": {
          "allow_unmapped_fields": false
        }
      }
    }    
  },
  "mappings": {
    "properties": {
      "query": {                
        "type": "percolator"            
      }
    }
  }
}'
```

2. Send a percolator document creation request or search request with a query containing a field where mapping does not exist
```sh
curl --location --request PUT 'localhost:9200/test/_doc/test-id' \
--header 'Content-Type: application/json' \
--data '{
  "query": {
    "query_string": {
      "query": "unmappedField: test"
    }
  }
}'
```

or

```sh
curl --location 'http://localhost:9200/test/_search' \
--header 'Content-Type: application/json' \
--data '{ 
  "query": { 
    "query_string": { 
      "query": "unmappedField: test" 
    }
  }
}'
```

- when query type is field query(when `getFieldQuery(String field, String queryText, int slop)`is called)
  - returns response with message that field mapping does not exist which is expected when `allow_unmapped_fields` is set to false
    - which is expected
    
```sh
{
  "error": {
    "root_cause": [
      {
        "type": "query_shard_exception",
        "reason": "No field mapping can be found for the field with name [unmappedField]",
        "index": "test",
        "index_uuid": "FSnBe-D3TEC8l1CQlu6uBQ"
      }
    ],
    "type": "search_phase_execution_exception",
    "reason": "all shards failed",
    "phase": "query",
    "grouped": true,
    "failed_shards": [
      {
        "shard": 0,
        "index": "test",
        "node": "U3GpcjP1RleoYleHB1NgSQ",
        "reason": {
          "type": "query_shard_exception",
          "reason": "No field mapping can be found for the field with name [unmappedField]",
          "index": "test",
          "index_uuid": "FSnBe-D3TEC8l1CQlu6uBQ"
        }
      }
    ]
  },
  "status": 400
}
```
- If the query type is one of the following:
  - wildcard query
    - `unmappedField: *` or `unmappedField: *test`
  - `_exists_` query
    - `_exists_: unmappedField`
  - range query
    - `unmappedField: <100`
  - fuzzy query 
    - `unmappedField: test~`
  - prefix query
    - `unmappedField: test*`
- 2xx response is returned
  - Internally parsed to `MatchNoDocsQuery` type, so it does not match any documents

### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [ ] API changes companion pull request [created](https://github.com/opensearch-project/opensearch-api-specification/blob/main/DEVELOPER_GUIDE.md).
- [ ] Failing checks are inspected and point to the corresponding known issue(s) (See: [Troubleshooting Failing Builds](../blob/main/CONTRIBUTING.md#troubleshooting-failing-builds))
- [ ] Commits are signed per the DCO using --signoff
- [ ] Commit changes are listed out in CHANGELOG.md file (See: [Changelog](../blob/main/CONTRIBUTING.md#changelog))
- [ ] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/index/search/QueryParserHelper.java [hunk 1]
```diff
@@ -143,7 +143,7 @@ static Map<String, Float> resolveMappingField(
                 fieldName = fieldName + fieldSuffix;
             }
 
-            MappedFieldType fieldType = context.getMapperService().fieldType(fieldName);
+            MappedFieldType fieldType = context.fieldMapper(fieldName);
```
[the remark this answers is not in the document]
Comment (imyp92) [reply server/src/main/java/org/opensearch/index/search/QueryParserHelper.java:146]: I've added yaml test cases and change log for this PR

**Comment by dblock** [comment]
@imyp92 Thanks! Amend your commits with `-s` for DCO, iterate to green.

**Comment by imyp92** [comment]
> @imyp92 Thanks! Amend your commits with `-s` for DCO, iterate to green.

I've amended all commits

**Review by gaobinlong — APPROVED** [review]

**Comment by dblock** [comment]
@gaobinlong please do hit merge when you approve so we don't get a conflict and have to restart gradles

**Comment by gaobinlong** [comment]
> @gaobinlong please do hit merge when you approve so we don't get a conflict and have to restart gradles

Sure, thank you!

**Pull request merged**